File: 22. Generate Parentheses/22. Generate Parentheses.dart
class Solution {
  List<String> generateParenthesis(int n) {
    List<String> result = [];
    backtrack(result, '', 0, 0, n);
    return result;
  }
  
  void backtrack(List<String> result, String current, int open, int close, int max) {
    if (current.length == max * 2) {
      result.add(current);
      return;
    }
    
    if (open < max) {
      backtrack(result, current + '(', open + 1, close, max);
    }
    if (close < open) {
      backtrack(result, current + ')', open, close + 1, max);
    }
  }
}

File: 22. Generate Parentheses/22. Generate Parentheses.rs
impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut result = Vec::new();
        Self::backtrack(&mut result, String::new(), 0, 0, n);
        result
    }
    
    fn backtrack(result: &mut Vec<String>, current: String, open: i32, close: i32, max: i32) {
        if current.len() == (max * 2) as usize {
            result.push(current);
            return;
        }
        
        if open < max {
            Self::backtrack(result, current.clone() + "(", open + 1, close, max);
        }
        if close < open {
            Self::backtrack(result, current + ")", open, close + 1, max);
        }
    }
}